question/classes/local/bank/column_base.php

```diff
@@ -0,0 +1,382 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Base class for representing a column.
+ *
+ * @package   core_question
+ * @copyright 1999 onwards Martin Dougiamas and others {@link http://moodle.com}
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace core_question\local\bank;
+
+defined('MOODLE_INTERNAL') || die();
+
+/**
+ * Base class for representing a column.
+ *
+ * @copyright 2009 Tim Hunt
+ * @author    2021 Safat Shahin <safatshahin@catalyst-au.net>
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+abstract class column_base {
+
+    /**
+     * @var view $qbank the question bank view we are helping to render.
+     */
+    protected $qbank;
+
+    /** @var bool determine whether the column is td or th. */
+    protected $isheading = false;
+
+    /**
+     * Constructor.
+     * @param view $qbank the question bank view we are helping to render.
+     */
+    public function __construct(view $qbank) {
+        $this->qbank = $qbank;
+        $this->init();
+    }
+
+    /**
+     * A chance for subclasses to initialise themselves, for example to load lang strings,
+     * without having to override the constructor.
+     */
+    protected function init(): void {
+    }
+
+    /**
+     * Set the column as heading
+     */
+    public function set_as_heading(): void {
+        $this->isheading = true;
+    }
+
+    /**
+     * Check if the column is an extra row of not.
+     */
+    public function is_extra_row(): bool {
+        return false;
+    }
+
+    /**
+     * Check if the row has an extra preference to view/hide.
+     */
+    public function has_preference(): bool {
+        return false;
+    }
+
+    /**
+     * Get if the preference key of the row.
+     */
+    public function get_preference_key(): string {
+        return '';
+    }
+
+    /**
+     * Get if the preference of the row.
+     */
+    public function get_preference(): bool {
+        return false;
+    }
+
+    /**
+     * Output the column header cell.
+     */
+    public function display_header(): void {
```

Comment:
I know this function I've called out that we should be using a renderer to generate the HTML snippets rather than just hard coding them.
I do wonder if this actually use a mustache template instead? It would be worth looking at

Reply:
created a renderer for this function